lib/Sema/CSSimplify.cpp

```diff
@@ -1091,14 +1091,19 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
                                      ConstraintLocatorBuilder locator) {
   // An @autoclosure function type can be a subtype of a
   // non-@autoclosure function type.
-  if (func1->isAutoClosure() != func2->isAutoClosure() &&
-      kind < ConstraintKind::Subtype)
-    return SolutionKind::Error;
+  if (func1->isAutoClosure() != func2->isAutoClosure()) {
+    // If the 2nd type is an autoclosure, then the first type needs wrapping in a
+    // closure despite already being a function type.
+    if (func2->isAutoClosure())
```

Comment:
I think this is actually a "functionality" change which requires test-case, because before we'd fall-through here if constraint was `>= Subtype` even if left-hand side wasn't an auto-closure, could you please add an example which captures this change in behavior?

Reply:
Since you are already changing this - it might be nice to unify `isAutoClosure`, `throws` and `isNoEscape` behavior into one function which compares attributes, because the logic is the same of all three...

Reply:
This code was refined after feedback from @CodaFi in #12040 (unmerged and closed). There is no functionality change because matchFunctionTypes() is bypassed by the sole caller when func1->isAutoClosure() is false and func2->isAutoClosure() is true. But as @CodaFi points out, the caller should stop bypassing matchFunctionTypes() and just handle the error.

Separately, I agree that the ExtInfo logic (in general) is messy and error prone. Not just here, but everywhere. It needs consolidating and clean up, but that is a separable problem.

Reply:
It’s good that it’s bypassed right now, but we should still have a test-case in case that ever changes...

Reply:
If the bypass is removed and unhandled, then the test suite fails:

```
test/Constraints/closures.swift:264:33: error: unexpected error produced: cannot convert value of type '((String) -> String)?' to expected argument type '_?'
    let _: (String) -> String = foo ?? bar
                                ^
```

Reply:
Right, but that relies on the fact that `??` currently has `<T>(T?, @autoclosure @autoclosure () throws -> T) throws -> T` overload, I think we might want to have a dedicated test-case for this, just in case.